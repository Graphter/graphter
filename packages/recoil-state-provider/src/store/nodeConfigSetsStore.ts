import { atom, RecoilState } from "recoil";
import { NodeConfig, PathSegment } from "@graphter/core";
import { pathToKey } from "@graphter/renderer-react";

export interface NodeConfigSets {
  activeConfigsKey: string,
  configSets: Map<string, Array<NodeConfig>>
}

const pathNodeConfigMap = new Map<string, RecoilState<NodeConfigSets>>()

const getPathKey = (path: Array<PathSegment>) => `node-configs-store-a12b32f8-ecb8-4522-bc9a-f1ab8a59caae-${pathToKey(path)}`

export const get = (
  path: Array<PathSegment>
): RecoilState<NodeConfigSets> => {
  checkPathArg(path)
  const key = getPathKey(path)
  if(!pathNodeConfigMap.has(key)) throw new Error(`Couldn't find configs for node at '${path.join('/')}'`)
  const state = pathNodeConfigMap.get(key)
  if(typeof state === 'undefined') throw new Error('Should not happen')
  return state
}

export const set = (
  path: Array<PathSegment>,
  configSets: NodeConfigSets
) => {
  checkPathArg(path)
  const pathKey = getPathKey(path)
  if(pathNodeConfigMap.has(pathKey)) throw new Error(`A value is already set for '${path.join('/')}'`)
  pathNodeConfigMap.set(pathKey, atom({
    key: pathKey,
    default: configSets
  }))
}

export const has = (path: Array<string | number>): boolean => {
  checkPathArg(path)
  return pathNodeConfigMap.has(getPathKey(path))
}

export const remove = (path: Array<PathSegment>) => {
  checkPathArg(path)
  pathNodeConfigMap.delete(getPathKey(path))
}


function checkPathArg(path: Array<PathSegment>){
  if(!path || !path.length) throw new Error('Path of at least one segment is required')
}

export const nodeConfigSetsStore = {
  get,
  set,
  has,
  remove
}