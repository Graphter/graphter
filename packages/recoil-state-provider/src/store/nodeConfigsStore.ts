import { atom, RecoilState } from "recoil";
import { NodeConfig, PathSegment } from "@graphter/core";
import { pathToKey } from "@graphter/renderer-react";

const pathNodeConfigMap = new Map<string, RecoilState<Array<NodeConfig>>>()

const getKey = (path: Array<PathSegment>) => `node-configs-store-a12b32f8-ecb8-4522-bc9a-f1ab8a59caae-${pathToKey(path)}`

export const get = (
  path: Array<PathSegment>
): RecoilState<Array<NodeConfig>> | null => {
  checkPathArg(path)
  const key = getKey(path)
  if(!pathNodeConfigMap.has(key)) return null
  const state = pathNodeConfigMap.get(key)
  if(typeof state === 'undefined') throw new Error('Should not happen')
  return state
}

export const set = (
  path: Array<PathSegment>,
  configs: Array<NodeConfig>
) => {
  console.log(`Setting configs for path '${path.join('/')}' to ${JSON.stringify(configs)}`)
  checkPathArg(path)

  const pathKey = getKey(path)

  if(pathNodeConfigMap.has(pathKey)) throw new Error(`A value is already set for '${path.join('/')}'`)

  pathNodeConfigMap.set(pathKey, atom({
    key: pathKey,
    default: configs
  }))
}

export const has = (path: Array<string | number>): boolean => {
  checkPathArg(path)
  return pathNodeConfigMap.has(getKey(path))
}

export const remove = (path: Array<PathSegment>) => {
  checkPathArg(path)
  pathNodeConfigMap.delete(getKey(path))
}


function checkPathArg(path: Array<PathSegment>){
  if(!path || !path.length) throw new Error('Path of at least one segment is required')
}

export const nodeConfigsStore = {
  get,
  set,
  has,
  remove
}