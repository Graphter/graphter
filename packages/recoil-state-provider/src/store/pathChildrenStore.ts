import { atom, RecoilState } from "recoil";
import { PathSegment } from "@graphter/core";
import { pathToKey } from "@graphter/renderer-react";

const childPathsMap = new Map<string, RecoilState<Array<any>>>()

const getKey = (path: Array<PathSegment>) => `path-children-store-a4d28666-5de8-4825-bcda-b1b6d116abe8-${pathToKey(path)}`

export const get = (
  path: Array<PathSegment>
): RecoilState<Array<any>> => {
  const key = getKey(path)
  if(!childPathsMap.has(key)) throw new Error(`Couldn't find path children state at '${path.join('/')}'`)
  const state = childPathsMap.get(key)
  if(typeof state === 'undefined') throw new Error('Should not happen')
  return state
}

export const set = (
  parentPath: Array<PathSegment>,
  childPaths: Array<Array<PathSegment>>
) => {

  const pathKey = getKey(parentPath)

  if(childPathsMap.has(pathKey)) throw new Error(`Tree path state already exists for '${parentPath.join('/')}'`)

  childPathsMap.set(pathKey, atom({
    key: pathKey,
    default: childPaths
  }))
}

export const has = (parentPath: Array<string | number>): boolean => {
  return childPathsMap.has(getKey(parentPath))
}

export const remove = (parentPath: Array<PathSegment>) => {
  childPathsMap.delete(getKey(parentPath))
}

export const pathChildrenStore = {
  get,
  set,
  has,
  remove
}