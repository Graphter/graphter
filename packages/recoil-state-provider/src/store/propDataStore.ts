/***
 * TODO: This module makes some pretty nasty data mutations. Fix that.
 */

import { atom, RecoilState } from "recoil";
import { PathSegment } from "@graphter/core";
import { nanoid } from "nanoid";

interface PropDataStateMeta {
  state: RecoilState<any>,
  deleted: boolean,
  committed: boolean,
  default: any
}

interface Node {
  id: string | null,
  meta: PropDataStateMeta | null,
  children: Array<Node> | null
}

const propStateNodeTree: Array<Node> = []

export const init = (
  path: Array<PathSegment>,
  data: any
): void => {
  const init = (path: Array<PathSegment>, data: any) => {
    if(Array.isArray(data)) {
      if(!has(path)) set(path, true, data.map(() => nanoid()))
      data.forEach((item, i) => init([ ...path, i ], item))
    } else if(typeof data === 'object') {
      if(!has(path)) set(path, true, data)
      Object.entries(data).forEach(([key, value]) => {
        init([ ...path, key ], value)
      })
    } else {
      if(!has(path)) set(path, true, data)
    }
  }
  init(path, data)
}

export const get = (
  path: Array<PathSegment>
): RecoilState<any> => {
  checkPathArg(path)
  const pathNode = getPathNode(path)
  if(!pathNode.meta) throw new Error(`Couldn't find state at ${path.join('/')}`)
  return pathNode.meta.state
}

export const getAll = (
  path: Array<PathSegment>
): Array<RecoilState<any>> => {
  checkPathArg(path)
  const pathNode = getPathNode(path)
  if(!pathNode.children) throw new Error(`Couldn't find child state at ${path.join('/')}`)
  return Array.from(pathNode.children.values())
    .filter(node => node.meta && node.meta.committed)
    .map(node => node.meta?.state as RecoilState<any>)
}

export const set = (
  path: Array<PathSegment>,
  committed: boolean,
  originalValue?: any
) => {
  checkPathArg(path)
  console.info(`Setting ${path.join('/')} = ${originalValue} [committed=${committed}]`)
  const node = getPathNode(path)
  node.meta = {
    state: atom({
      key: nanoid(),
      default: originalValue
    }),
    deleted: false,
    committed,
    default: originalValue
  }
}

/***
 * The "committed" flag in Graphter is used to demarcate draft tree additions. For example
 * the "add item" ui for a list before the "Add [+]" button is pressed.
 *
 * In future we'll likely have to integrate this function with node modification history or
 * versioning to allow us to track tree modifications as well as additions. For example a
 * renderer may want to allow changes and then provide a Save/Cancel button to commit or
 * discard the tree changes
 * @param path
 */
export const commitItem = (
  path: Array<PathSegment>
) => {
  checkPathArg(path)
  const node = getPathNode(path)
  if(!node.meta) throw new Error(`Node at ${path.join('/')} is missing meta. Should never happen`)
  node.meta.committed = true
}

export const has = (path: Array<string | number>): boolean => {
  checkPathArg(path)
  let node: Node | undefined = {
    id: null,
    meta: null,
    children: propStateNodeTree
  }
  for(let segment of path){
    if(typeof segment === 'string'){
      node = node.children?.find(childNode => childNode.id === segment)
    } else {
      node = node.children ? node.children[segment] : undefined
    }

    if(!node) return false
  }
  return true
}

export const remove = (path: Array<PathSegment>) => {
  checkPathArg(path)
  const clonedPath = [...path]
  const removeSegment = clonedPath.splice(-1, 1)[0]

  const parentNode = getPathNode(clonedPath)
  if(!parentNode) throw new Error(`Missing ancestor trying to set ${path.join('/')}`)
  if(!parentNode.children) throw new Error(`No children at ${path.join('/')} to remove`)
  const indexToRemove = typeof removeSegment === 'number' ?
    removeSegment :
    parentNode.children
      .findIndex(childNode => childNode.id !== removeSegment)
  parentNode.children.splice(indexToRemove, 1)
}

const getPathNode = (path: Array<PathSegment>) => {
  return path.reduce<Node>((a, c) => {
    if(!a.children) a.children = []
    let node = typeof c === 'string' ?
      a.children.find(node => node.id === c) :
      a.children[c]

    if(!node){
      node = {
        id: typeof c === 'string' ? c : null,
        meta: null,
        children: null
      }
      typeof c === 'string' ?
        a.children.push(node) :
        a.children[c] = node
    }
    return node
  }, { id: null, meta: null, children: propStateNodeTree })
}

function checkPathArg(path: Array<PathSegment>){
  if(!path || !path.length) throw new Error('Path of at least one segment is required')
}

export interface PropDataStore {
  init: (path: Array<PathSegment>, data: any) => void
  get: (path: Array<PathSegment>) => RecoilState<any>
  getAll: (path: Array<PathSegment>) => Array<RecoilState<any>>
  set: (
    path: Array<PathSegment>,
    committed: boolean,
    originalValue?: any
  ) => void
  commitItem: (path: Array<PathSegment>) => void
  has: (path: Array<string | number>) => boolean
  remove: (path: Array<PathSegment>) => void
}

export const propDataStore: PropDataStore = {
  init,
  get,
  getAll,
  set,
  commitItem,
  has,
  remove
}

export default PropDataStore