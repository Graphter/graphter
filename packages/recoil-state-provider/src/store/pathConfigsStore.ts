import { selector, RecoilValueReadOnly } from "recoil";
import { NodeConfig, PathSegment } from "@graphter/core";
import { pathToKey } from "@graphter/renderer-react";
import { getPathPaths } from "../utils/getPathPaths";
import { nodeConfigSetsStore } from "./nodeConfigSetsStore";

const pathConfigsMap = new Map<string, RecoilValueReadOnly<Array<Array<NodeConfig>>>>()

export const get = (
  path: Array<PathSegment>
): RecoilValueReadOnly<Array<Array<NodeConfig>>> => {
  const pathKey = `path-config-store-${pathToKey(path)}`
  let pathConfigsSelector = pathConfigsMap.get(pathKey)
  if(pathConfigsSelector) return pathConfigsSelector
  pathConfigsSelector = selector({
    key: pathKey,
    get: ({ get }) => {
      const pathPaths = getPathPaths(path)
      return pathPaths.reduce<Array<Array<NodeConfig>>>((a, c) => {
        const nodeConfigSetsState = nodeConfigSetsStore.get(c)
        const nodeConfigSets = get(nodeConfigSetsState)
        const configs = nodeConfigSets.configSets.get(nodeConfigSets.activeConfigsKey)
        if (configs) a.push(configs)
        return a
      }, [])
    }
  })
  pathConfigsMap.set(pathKey, pathConfigsSelector)
  return pathConfigsSelector
}

export const pathConfigsStore = {
  get,
}