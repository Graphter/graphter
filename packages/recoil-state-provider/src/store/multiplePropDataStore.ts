import { RecoilValueReadOnly, selector } from "recoil";
import { PathSegment } from "@graphter/core";
import { propDataStore } from "./propDataStore";

const multiplePropDataMap: Map<string, RecoilValueReadOnly<any>> = new Map()
const keySalt = '5042b962-fd6c-470c-bbae-747380814365'
const get = (paths: Array<Array<PathSegment>>): RecoilValueReadOnly<Array<{ path: Array<PathSegment>, data: any }>> => {
  const key = paths.map(path => path.join(keySalt)).join('-')
  if(!multiplePropDataMap.has(key)){
    multiplePropDataMap.set(key, selector<Array<{ path: Array<PathSegment>, data: any }>>({
      key: key,
      get: ({ get }) => {
        return paths.map(path => {
          const state = propDataStore.get(path)
          if(!state) throw new Error(`Missing state for ${path.join('/')}'`)
          const pathMeta = get(state)
          return {
            path,
            data: pathMeta
          }
        })
      }
    }))
  }
  const state = multiplePropDataMap.get(key)
  if(!state) throw new Error(`Error gathering state for paths: ${paths.map(path => path.join('/')).join(' - ')}`)
  return state
}

const multiplePropDataStore = {
  get
}

export default multiplePropDataStore