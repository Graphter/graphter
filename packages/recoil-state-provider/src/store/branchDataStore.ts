import { RecoilValueReadOnly, selector } from "recoil";
import { PathSegment } from "@graphter/core";
import { nodeConfigSetsStore } from "./nodeConfigSetsStore";
import { pathChildrenStore } from "./pathChildrenStore";
import { nodeRendererStore } from "@graphter/renderer-react";
import { rendererInternalDataStore } from "./rendererInternalDataStore";
import { pathToKey } from "@graphter/renderer-react";
import { pathConfigsStore } from "./pathConfigsStore";
import { getExactPathConfigs } from "../utils/getExactPathConfigs";
import { pathConfigsToString } from "../utils/pathConfigsToString";

const treeDataMap: Map<string, RecoilValueReadOnly<any>> = new Map()

const get = <T>(startingPath: Array<PathSegment>, depth?: number) => {
  if (!startingPath) throw new Error('Path is required to get descendent data')
  const key = pathToKey(startingPath)
  let treeDataSelector = treeDataMap.get(key)
  if (treeDataSelector) return treeDataSelector

  treeDataSelector = selector<T>({
    key: key,
    get: ({get}) => {
      function getNodeData(path: Array<PathSegment>): any {
        const nodeConfigSetsState = nodeConfigSetsStore.get(path)
        if (!nodeConfigSetsState) throw new Error(`Missing path config state at path ${path.join('/')}`)
        const nodeConfigSets = get(nodeConfigSetsState)
        const nodeConfigs = nodeConfigSets.configSets.get(nodeConfigSets.activeConfigsKey)
        if(!nodeConfigs) throw new Error(`Can't find active configs at '${path.join('/')}'`)

        function getChildData(){
          if(!pathChildrenStore.has(path)) return null
          const childPathsState = pathChildrenStore.get(path)
          if (!childPathsState) return null
          const childPaths = get<Array<Array<PathSegment>>>(childPathsState)
          return childPaths.length ?
            childPaths.map(childPath => getNodeData(childPath)) :
            null
        }

        const childData = getChildData()
        const pathConfigsState = pathConfigsStore.get(path)
        const pathConfigs = get(pathConfigsState)
        // transform from the bottom most path node -> up
        const externalNodeData = [ ...nodeConfigs ].reverse().reduce<any>((a, c) => {
          const exactPathConfigs = getExactPathConfigs(pathConfigs, c)
          const internalDataState = rendererInternalDataStore.get(path, exactPathConfigs)
          if (!internalDataState){
            throw new Error(`Missing internal data for ${
              pathConfigsToString(exactPathConfigs)
            } at path '${path.join('/')}' node ${c.id}`)
          }
          const internalData = get(internalDataState)
          const rendererReg = nodeRendererStore.get(c.type)
          const rendererInternalData = typeof internalData !== 'undefined' ? internalData : a
          if (!childData?.length || !rendererReg.mergeChildData) return rendererInternalData
          else return rendererReg.mergeChildData(c, path, rendererInternalData, getNodeData, childData)
        }, undefined)

        return externalNodeData
      }

      return getNodeData(startingPath)
    }
  });
  treeDataMap.set(key, treeDataSelector)

  return treeDataSelector
}

const has = (startingPath: Array<PathSegment>) => {
  if (!startingPath) throw new Error('Path is required to get descendent data')
  const key = pathToKey(startingPath)
  return treeDataMap.has(key)
}

export const branchDataStore = {
  get,
  has
}

export default branchDataStore