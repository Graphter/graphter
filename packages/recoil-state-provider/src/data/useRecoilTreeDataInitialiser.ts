import { propDataStore } from "../store/propDataStore";
import { TreeDataInitialiserHook } from "@graphter/renderer-react";
import { getTreeMeta } from "@graphter/renderer-react";
import { PathSegment } from "@graphter/core";
import { getValue } from "@graphter/renderer-react"
import { createDefault, nodeRendererStore } from "@graphter/renderer-react";

const NO_MATCH = 'no-match-ac437190-4001-4fe3-bf1f-8ad03366d1dd'

export const useRecoilTreeDataInitialiser: TreeDataInitialiserHook = () => {
  return (config, path, committed = true, originalTreeData) => {

    const getNodeValue = (path: Array<PathSegment>) => {
      const result = getValue(originalTreeData, path, NO_MATCH)
      if (result === NO_MATCH) return
      return result
    }

    const localPath = path.slice(2)
    const treeMeta = getTreeMeta(config, localPath, getNodeValue)

    for (const nodeMeta of treeMeta) {
      if (!nodeMeta.path.length) continue
      const globalPath = [ ...path.slice(0, 2), ...nodeMeta.path ]
      if (propDataStore.has(globalPath)) continue
      const rendererReg = nodeRendererStore.get(nodeMeta.config.type)
      if(rendererReg.initialiser){
        const initialData = rendererReg.initialiser(originalTreeData, nodeMeta.config, nodeMeta.path)
        propDataStore.set(globalPath, initialData)
      } else {
        let initialData = getValue(originalTreeData, nodeMeta.path, NO_MATCH)
        if (initialData === NO_MATCH) {
          initialData = createDefault(
            nodeMeta.config,
            rendererReg.createFallbackDefaultValue ?
              rendererReg.createFallbackDefaultValue(nodeMeta.config, globalPath, getNodeValue) :
              null
          )
        }
        propDataStore.set(globalPath, initialData)
      }



    }

  }

}