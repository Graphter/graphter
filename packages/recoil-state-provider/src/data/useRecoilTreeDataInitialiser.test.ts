import { propDataStore } from "../store/propDataStore";
import { nodeRendererStore } from "@graphter/renderer-react";
import { when } from "jest-when";
import { useRecoilTreeDataInitialiser } from "./useRecoilTreeDataInitialiser";

jest.mock('../store/propDataStore')

const nodeRendererStoreMock = nodeRendererStore as jest.Mocked<any>

describe('useRecoilTreeDataInitialiser()', () => {
  beforeEach(() => {
    jest.resetAllMocks()
  })
  it('should start initialising state from the top node downward', () => {
    const initMock = jest.fn()
    when(nodeRendererStoreMock.get)
      .calledWith('some-type')
      .mockReturnValueOnce({
        initialiseData: initMock
      })
    const init = useRecoilTreeDataInitialiser()
    init(
      { id: 'some-id', type: 'some-type' },
      [ 'page' ],
      { some: 'data' }
    )
    expect(initMock).toHaveBeenCalled()
    const params = initMock.mock.calls[0]
    expect(params).toMatchSnapshot()
    expect(propDataStore.set).not.toHaveBeenCalled()
    params[3](
      [ 'page', 'body' ],
      { someOther: 'data' }
    )
    expect(propDataStore.set).toHaveBeenCalledWith(
      [ 'page', 'body' ],
      true,
      { someOther: 'data' }
    )
  })
  it(`should throw if a renderer can't be found for the top node`, () => {
    nodeRendererStoreMock.get.mockReturnValueOnce(null)
    const init = useRecoilTreeDataInitialiser()
    expect(() => init(
      { id: 'some-id', type: 'some-type' },
      [ 'page' ],
      { some: 'data' }
    )).toThrowErrorMatchingSnapshot()
  })
  it(`should initialise the entire tree if the top node renderer registration doesn't have an init fn`, () => {
    when(nodeRendererStoreMock.get)
      .calledWith('some-type')
      .mockReturnValueOnce({ })
    const init = useRecoilTreeDataInitialiser()
    init(
      { id: 'some-id', type: 'some-type' },
      [ 'page' ],
      { some: 'data' }
    )
    expect(propDataStore.set).toHaveBeenCalledWith(
      [ 'page' ],
      true,
      { some: 'data' }
    )
  })
})
export {}