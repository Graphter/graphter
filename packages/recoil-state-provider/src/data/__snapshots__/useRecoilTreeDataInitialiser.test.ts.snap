// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useRecoilTreeDataInitialiser() should start initialising state from the top node downward 1`] = `
Array [
  Object {
    "id": "some-id",
    "type": "some-type",
  },
  Array [
    "page",
  ],
  Object {
    "some": "data",
  },
  [Function],
]
`;

exports[`useRecoilTreeDataInitialiser() should throw if a renderer can't be found for the top node 1`] = `"Unable to find renderer for node type 'some-type' at page"`;
